### PR TITLE
A faster copyto_unaliased!

### PR DESCRIPTION
**Pull request by N5N3**
~At present, `a .= b` falls to `copyto!(a, b)` if `axes(a) == axes(b)` for performance optimization.
However, as mentioned in #40962 and #39345, the performance of `copyto!(a, b)` is very low in many cases. Thus, I think we can implement a faster version here, based on `@simd`, to solve the problem to some extent.
(I tried to to optimize `copyto_unalias!`, but many generated codes are not vectorized without `@simd`.)~
This PR tries to speed up `copyto_unaliased!` in many cases:
1. **Linear-to-Linear copy** is using `Base.OneTo` as the iterartor instead of `LinearIndices` for faster speed.
2. `zip()` based iterator is used to speed up the most general case.
3. The innermost loop for `CartesianIndices` is separated mannully like `@simd` (but add no `Expr(:loopinfo`) to speed up if the 1st dim's length is larger than 16.
```julia
size:(16, 2048) -> size:(16, 2048)
Car -> Car : before: 17.5μs -> after: 8.5μs
Car -> Lin : before: 52.2μs -> after: 8.5μs
Lin -> Car : before: 75.5μs -> after: 8.77μs
Lin -> Lin : before: 10.6μs -> after: 8.3μs
size:(16, 4, 512) -> size:(16, 4, 512)
Car -> Car : before: 18μs -> after: 9.5μs
Car -> Lin : before: 50.1μs -> after: 9.3μs
Lin -> Car : before: 82μs -> after: 11.6μs
Lin -> Lin : before: 10.7μs -> after: 8.37μs
size:(16, 4, 4, 128) -> size:(16, 4, 4, 128)
Car -> Car : before: 19.7μs -> after: 9.4μs
Car -> Lin : before: 79.3μs -> after: 9.2μs
Lin -> Car : before: 95.8μs -> after: 9.2μs
Lin -> Lin : before: 10.6μs -> after: 8.37μs
size:(16, 4, 4, 4, 32) -> size:(16, 4, 4, 4, 32)
Car -> Car : before: 20.4μs -> after: 9.8μs
Car -> Lin : before: 94.9μs -> after: 9.1μs
Lin -> Car : before: 119μs -> after: 10μs
Lin -> Lin : before: 10.6μs -> after: 8.37μs
```
The above Benchmarks, using `SubArray{Float64}`, are done in 1.7.0-beta3, where the performance regression on `CartesianIndices`'s iteration(#38073) seems to be fixed.

### CODE REVIEW

**Comment by N5N3** [comment]
The failure in mpfr seems unrelated？

**Comment by oscardssmith** [comment]
Can you add a 3d benchmark? Other than that, this looks good.

**Review by oscardssmith — APPROVED** [review]

**Comment by N5N3** [comment]
It seems the speed of the general broadcast kernal is strongly affected by the array's shape.
50*50 is almost 5 times faster than 5*500.
Maybe the `exclude` in `preprocess` should be omitted in this case for better performance?

**Comment by oscardssmith** [comment]
Why did you close this?

**Comment by N5N3** [comment]
Just for reverting the branch, and try to speed up `copyto_unalias!`.
The acceleration is considerable on my Desktop, but the code might be unclear (include some `@simd` that should not appear)

**Comment by N5N3** [comment]
Opps, it seems we can't use `@simd` here. But it does speed up a lot.

File: base/broadcast.jl [hunk 1]
```diff
@@ -991,11 +991,17 @@ preprocess_args(dest, args::Tuple{}) = ()
 # Specialize this method if all you want to do is specialize on typeof(dest)
 @inline function copyto!(dest::AbstractArray, bc::Broadcasted{Nothing})
     axes(dest) == axes(bc) || throwdm(axes(dest), axes(bc))
-    # Performance optimization: broadcast!(identity, dest, A) is equivalent to copyto!(dest, A) if indices match
+    # Performance optimization: broadcast!(identity, dest, A) is equivalent to copyto!(dest, A) if indices match.
+    # However copyto!(dest, A) is very slow in many cases, implement a faster version here.
```
Comment (stevengj) [inline base/broadcast.jl:995]: Wouldn't it be better to speed up `copyto!(dest, A)`?

Comment (N5N3) [reply base/broadcast.jl:995]: In fact, I tried, but the problem seems to be a little complicated with the following facts:

1. `copyto!(::Array,::Array)` call C's `memmove`, but it seems slower than a single loop:
```julia
julia> @btime copyto!($b,$a);                    # this call C's memmove
  120.524 ns (0 allocations: 0 bytes)
julia> @btime copyto!(IndexLinear(),$b,IndexLinear(),$a);  # this call Julia's copyto_unalias!
  74.512 ns (0 allocations: 0 bytes)
```
2.  For `abstractarray`, I can't find a way to accelerate `copyto_unalias!(dest, A)` without `@simd`, as:
```julia
a = randn(40,40); b = similar(a);
f1!(a,b) = @inbounds for i in eachindex(IndexCartesian(), a)
    b[i] = a[i]
end

f2!(a,b) = @inbounds @simd for i in eachindex(IndexCartesian(), a)
    b[i] = a[i]
end
@btime f1!($a, $b)   #247.132 ns (0 allocations: 0 bytes)
@btime f2!($a, $b)  #220.548 ns (0 allocations: 0 bytes)  a little faster
@btime @views f1!($a[1:end,1:end], $b[1:end,1:end]) #   1.710 μs (0 allocations: 0 bytes)
@btime @views f2!($a[1:end,1:end], $b[1:end,1:end]) #   262.500 ns (0 allocations: 0 bytes) much faster
```

Comment (stevengj) [reply base/broadcast.jl:995]: Why can't you use `@simd`?

Comment (N5N3) [reply base/broadcast.jl:995]: The buildbot throw an error that ’@simd‘ is not defined. I‘m not familiar with the build system, but it seems to be a world age problem?(abstractarray.jl is included before simd.jl in Bass.jl)

Comment (stevengj) [reply base/broadcast.jl:995]: Sounds just like a bootstrapping problem, i.e. `@simd` is not defined until later in the build process.  That should be fixable by e.g. by moving the `copyto!` definition to a file loaded later.

**Comment by mcabbott** [comment]
See also #38014, a less elaborate attempt which ran into I think the same bootstrap problem. 

There were concerns raised there about arrays like `BitArray` which alias themselves; I don't know whether or not this PR avoids that.

**Comment by KristofferC** [comment]
> There were concerns raised there about arrays like BitArray which alias themselves; 

That's only relevant if one uses the `ivdep` though? I'm a bit surprised why `@simd` should have any effect in copying. I don't think it should give any extra leeway to the compiler so not sure why it would improve performance.

**Comment by stevengj** [comment]
Yes, I can't see any performance difference from `@simd` in something like:
```jl
function foo!(dest, src)
    for i in eachindex(src)
        @inbounds dest[i] = src[i]
    end
end
```
which already uses `vmovups` according to `@code_native`.   What is the case that benefits from `@simd`?

**Comment by vtjnash** [comment]
IIRC, the `@simd` macro also tries to rearrange some loops to attempt to make them more linear and more amenable to analysis, though you can get the same (or better?) just by not forcing it to use `IndexCartesian()`

**Comment by N5N3** [comment]
I hope this is reproducable @stevengj 
```julia
function simdfoo!(dest, src)
    @inbounds @simd for i in eachindex(src)
        dest[i] = src[i]
    end
end
function foo!(dest, src)
    @inbounds for i in eachindex(src)
        dest[i] = src[i]
    end
end

a = randn(8,8,8,8);
a = @view a[axes(a)...];
b = similar(a);
b = @view b[axes(a)...];
@btime foo!($b, $a) #  6.480 μs (0 allocations: 0 bytes)
@btime simdfoo!($b, $a) #  2.000 μs (0 allocations: 0 bytes)
```

**Comment by KristofferC** [comment]
I think this is https://github.com/JuliaLang/julia/issues/38073 cc @kimikage 

On 1.5:

```
julia> @btime foo!($b, $a) #  6.480 μs (0 allocations: 0 bytes)
  1.658 μs (0 allocations: 0 bytes)

julia> @btime simdfoo!($b, $a) #  2.000 μs (0 allocations: 0 bytes)
  2.368 μs (0 allocations: 0 bytes)
```

On 1.6:

```
julia> @btime foo!($b, $a) #  6.480 μs (0 allocations: 0 bytes)
  6.112 μs (0 allocations: 0 bytes)

julia> @btime simdfoo!($b, $a) #  2.000 μs (0 allocations: 0 bytes)
  1.768 μs (0 allocations: 0 bytes)
```

**Comment by N5N3** [comment]
Some more benchmarks, It seems 1.6.1 is stably slower than 1.5.4 even with a `@simd` on 4，5，6D `CartesianIndices`'s iteration .
```julia
julia> a = randn(4,4,4,4,4,4); b = similar(a); a = view(a,axes(a)...); b = view(b,axes(b)...);
julia> @btime foo!($b,$a);            #  1.6.1: 9.200 μs      1.5.4: 3.112 μs
julia> @btime simdfoo!($b,$a);        #  1.6.1: 3.763 μs      1.5.4: 6.040 μs

julia> a = randn(5,5,5,5,5); b = similar(a); a = view(a,axes(a)...); b = view(b,axes(b)...);
julia> @btime foo!($b,$a);            #  1.6.1: 6.000 μs      1.5.4: 2.144 μs
julia> @btime simdfoo!($b,$a);        #  1.6.1: 2.378 μs      1.5.4: 3.287 μs

julia> a = randn(8,8,8,8); b = similar(a); a = view(a,axes(a)...); b = view(b,axes(b)...);
julia> @btime foo!($b,$a);            #  1.6.1: 6.420 μs      1.5.4: 1.840 μs
julia> @btime simdfoo!($b,$a);        #  1.6.1: 1.990 μs      1.5.4: 2.678 μs

julia> a = randn(16,16,16); b = similar(a); a = view(a,axes(a)...); b = view(b,axes(b)...);
julia> @btime foo!($b,$a);            #  1.6.1: 5.833 μs      1.5.4: 961.857 ns
julia> @btime simdfoo!($b,$a);        #  1.6.1: 747.619 ns    1.5.4: 999.900 ns

julia> a = randn(64,64); b = similar(a); a = view(a,axes(a)...); b = view(b,axes(b)...);
julia> @btime foo!($b,$a);            #  1.6.1: 3.737 μs      1.5.4: 5.650 μs
julia> @btime simdfoo!($b,$a);        #  1.6.1: 512.500 ns    1.5.4: 699.306 ns
```

**Comment by N5N3** [comment]
@mcabbott 
> There were concerns raised there about arrays like `BitArray` which alias themselves; I don't know whether or not this PR avoids that. (E.g. I don't know whether messing with `Expr(:loopinfo` is identical to `@simd`.)

I tried to remove every `Expr(:loopinfo` generated by `@simd` while keeping the loop body.
The results shows that the acceleration is not influenced. 
Since the expanded code is equivalent, I think this is safe enough?
Maybe we can add a safe option for `@simd` like:
```julia
@simd safe for i in 1:10
end
```
which only expands the loop but not marks it as SIMD.

**Pull request closed without merging**